### PR TITLE
Add reshape op supported by MKL-DNN

### PR DESCRIPTION
**Pull request by huangzhiyuan**
## Description ##
This PR indents to add reshape op supported by MKL-DNN. It will help reduce the unnecessary reorder operations after a MKL-DNN operation. @pengzhao-intel @ZhennanQin

***Test network:***
```
data = mx.symbol.Variable('data')
conv1 = mx.symbol.Convolution(data=data, num_filter=16384, kernel=(1, 1), pad=(0, 0), stride=(1, 1))
res = mx.symbol.reshape(data=conv1, shape=(16384,4,16,16))
trans = mx.symbol.transpose(data=res)
res1 = mx.symbol.reshape(data=trans, shape=(1, 1024, 1, 16384))
conv2 = mx.symbol.Convolution(data=res1, num_filter=16384, kernel=(1, 1), pad=(0,0), stride=(1,1))
```

**Origin verbose log:**

mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nchw fwei:Ohwi16o fbia:x fdst:nChw16c,alg:convolution_direct,mb4_g1ic3oc16384_ih16oh16kh1sh1dh0ph0_iw16ow16kw1sw1dw0pw0,2.08691
**mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nChw16c out:f32_nchw,num:1,4x16384x16x16,2.44995**
**mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nChw16c,num:1,4x16384x16x16,1.64697**
**mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nChw16c out:f32_nchw,num:1,4x16384x16x16,2.30298**
**mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw out:f32_nchw,num:1,16384x4x16x16,2.1731**
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nChw16c,num:1,1x1024x1x16384,2.40698
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,16384x1024x1x1,2.29297
mkldnn_verbose,exec,convolution,jit_1x1:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:x fdst:nChw16c,alg:convolution_direct,mb1_g1ic1024oc16384_ih1oh1kh1sh1dh0ph0_iw16384ow16384kw1sw1dw0pw0,457.164

**After reshape op supported by MKL-DNN verbose log:**

mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nchw fwei:Ohwi16o fbia:x fdst:nChw16c,alg:convolution_direct,mb4_g1ic3oc16384_ih16oh16kh1sh1dh0ph0_iw16ow16kw1sw1dw0pw0,1.86304
**mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw16c out:f32_nchw,num:1,4x16384x16x16,2.35913**
**mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw out:f32_nchw,num:1,16x16x4x16384,1.55103**
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nchw out:f32_nChw16c,num:1,1x1024x1x16384,2.6958
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_oihw out:f32_OIhw16i16o,num:1,16384x1024x1x1,2.49683
mkldnn_verbose,exec,convolution,jit_1x1:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:x fdst:nChw16c,alg:convolution_direct,mb1_g1ic1024oc16384_ih1oh1kh1sh1dh0ph0_iw16384ow16384kw1sw1dw0pw0,432.984

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by ankkhedia** [comment]
@huangzhiyuan Thanks for the contribution!
Could you please make sure that the changes in the PR passes CI
@mxnet-label-bot [pr-awaiting-testing , MKLDNN, operator]

**Comment by ankkhedia** [comment]
@huangzhiyuan It seems that the PR is still failing on CI. Request you to take a look.

**Comment by huangzhiyuan** [comment]
@ankkhedia please take a view, ths :)

**Comment by ankkhedia** [comment]
@azai91 @mseth10 Could you please take a look into this PR?

**Comment by pengzhao-intel** [comment]
ping @zheng-da @azai91 @mseth10

**Comment by kalyc** [comment]
@mxnet-label-bot update [pr-awaiting-review]

**Comment by stu1130** [comment]
ping @azai91 @mseth10 to review

**Comment by vandanavk** [comment]
@mxnet-label-bot add [MKLDNN]

File: src/operator/tensor/matrix_op.cc [hunk 1]
```diff
@@ -103,6 +103,57 @@ DMLC_REGISTER_PARAMETER(StackParam);
...
+  CHECK_EQ(outputs.size(), 1U);
+  // If inputs are supposed to be in MKLDNN format and
+  // MKLDNNsupport the data type or the shape. Then convert
+  // it to the output format and shape
+  if (SupportMKLDNNArray(inputs[0].dtype(), inputs[0].shape()) && req[0] != kAddTo) {
+    MKLDNNReshape(inputs[0], outputs[0]);
+    return;
+  }
+  FallBackCompute(UnaryOp::IdentityCompute<cpu>, attrs, ctx, inputs, req,
+                    outputs);
```
Comment (ZhennanQin) [inline src/operator/tensor/matrix_op.cc]: Indent.

File: src/operator/tensor/matrix_op.cc [hunk 2]
```diff
@@ -261,7 +317,9 @@ Example::
 .set_num_outputs(1)
 .set_attr<nnvm::FInferShape>("FInferShape", FlattenShape)
 .set_attr<nnvm::FInferType>("FInferType", ElemwiseType<1, 1>)
+#if MXNET_USE_MKLDNN == 1
```
Comment (ZhennanQin) [inline src/operator/tensor/matrix_op.cc]: If MKLDNN supports this op, please add `.set_attr<bool>("TIsMKLDNN", true)` for it.

Comment (huangzhiyuan) [reply src/operator/tensor/matrix_op.cc]: This attribute is in Line 327.

**Comment by huangzhiyuan** [comment]
@azai91 @mseth10 could you help review the PR? This has been out for review for a long long time...

File: src/operator/tensor/matrix_op.cc [hunk 3]
```diff
@@ -210,24 +272,18 @@ static void FlattenEx(const nnvm::NodeAttrs& attrs,
 #endif
 }
 
+#if MXNET_USE_MKLDNN == 1
```
Comment (TaoLv) [inline src/operator/tensor/matrix_op.cc]: Seems this will change the original behavior when MKL-DNN is not enabled. `FlattenStorageType` is not defined then.

File: src/operator/tensor/matrix_op.cc [hunk 4]
```diff
@@ -914,7 +972,7 @@ NNVM_REGISTER_OP(depth_to_space)
 .describe(R"code(Rearranges(permutes) data from depth into blocks of spatial data.
 Similar to ONNX DepthToSpace operator:
 https://github.com/onnx/onnx/blob/master/docs/Operators.md#DepthToSpace.
-The output is a new tensor where the values from depth dimension are moved in spatial blocks
```
Comment (TaoLv) [inline src/operator/tensor/matrix_op.cc]: Thank you for fixing these white spaces at line-end. Does cpplint complain about them? If not, I would like to keep them as is since these changes are not related to this PR.

File: src/operator/tensor/matrix_op.cc (hunk 3 above)
Comment (ZhennanQin) [reply src/operator/tensor/matrix_op.cc]: This is followed new style with other mkldnn op, that is, defining InferStorageType function within MKLDNN  macro. Most other ops are refactored into this style by Luobao, I guess this one is missing.

File: src/operator/tensor/matrix_op.cc (hunk 4 above)
Comment (huangzhiyuan) [reply src/operator/tensor/matrix_op.cc]: Have fixed the white space same as before, could you help review again? thanks :)

**Comment by TaoLv** [comment]
Is there any unit test to cover this change?

**Comment by huangzhiyuan** [comment]
@TaoLv @azai91 @mseth10  I've made the change and unit test based on your suggestions. What do you think?

File: src/operator/tensor/matrix_op.cc [hunk 5]
```diff
@@ -171,9 +222,19 @@ If the argument `reverse` is set to 1, then the special values are inferred from
 .set_num_outputs(1)
 .set_attr_parser(ParamParser<ReshapeParam>)
 .set_attr<nnvm::FInferShape>("FInferShape", ReshapeShape)
+#if MXNET_USE_MKLDNN == 1
+.set_attr<FInferStorageType>("FInferStorageType", ReshapeStorageType)
+#endif
 .set_attr<nnvm::FInferType>("FInferType", ElemwiseType<1, 1>)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseNone{"_backward_copy"})
 .set_attr<FCompute>("FCompute<cpu>", UnaryOp::IdentityCompute<cpu>)
+#if MXNET_USE_MKLDNN == 1
```
Comment (anirudh2290) [inline src/operator/tensor/matrix_op.cc]: I know that this is not the first op which is implemented this way, but is there a reason for the two different if blocks here ?

Comment (huangzhiyuan) [reply src/operator/tensor/matrix_op.cc]: Nothing different, it has been merged into one block here. Thanks your review! :)
